File: src/gimelstudio/core/eval_info.py
# ----------------------------------------------------------------------------
# Gimel Studio Copyright 2019-2023 by the Gimel Studio project contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ----------------------------------------------------------------------------


class EvalInfo(object):
    """
    Evaluate node properties and parameters
    """
    def __init__(self, node):
        if node is None:
            raise TypeError
        self.node = node

    def EvaluateProperty(self, name):
        prop = self.node.properties[name]
        if prop.binding:
            # Evaluate the next node
            binding = prop.binding
            info = EvalInfo(binding[0])
            return binding[0].EvaluateNode(info)[binding[1]]
        return prop.value

File: lib/models/eval_info.dart
import 'package:gimelstudio/models/node_base.dart';
import 'package:gimelstudio/models/node_property.dart';

class EvalInfo {
  EvalInfo({required this.node});

  final NodeBase node;

  /// Evaluate the value of the parameter [name] of [node].
  dynamic evaluateProperty(String name) {
    Property prop = node.properties.values.firstWhere((item) => item.idname == name);
    if (prop.connection != null) {
      // Evaluate the next node
      NodeBase connection = prop.connection!.$1;
      String outputIdname = prop.connection!.$2;
      EvalInfo info = EvalInfo(node: connection);
      return connection.evaluateNode(info)[outputIdname];
    }
    return prop.value;
  }
}
